openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py

```diff
@@ -0,0 +1,210 @@
+    """
+    Free-text object tag used on system-defined taxonomies
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+    
+class ClosedSystemObjectTag(ClosedObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Object tags linked to a closed system-taxonomy
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+
+class ModelObjectTag(OpenSystemObjectTag):
+    """
+    Object tags used with tags that relate to the id of a model
+    
+    This object tag class is not registered as it needs to have an associated model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = None
+
+
+    def _check_taxonomy(self):
+        """
+        Validates if has an associated Django model that has an Id
+        """
+        if not super()._check_taxonomy():
+            return False
+    
+        if not self.tag_class_model:
+            return False
+
+        # Verify that is a Django model
+        if not issubclass(self.tag_class_model, models.Model):
+            return False
+
+        # Verify that the model has 'id' field
+        try:
+            self.tag_class_model._meta.get_field('id')
+        except FieldDoesNotExist:
+            return False
+
+        return True
+    
+    def _check_instance(self):
+        """
+        Validates if the instance exists
+        """
+        try:
+            intance_id = int(self.value)
+        except ValueError:
+            return False
+        return self.tag_class_model.objects.filter(id=intance_id).exists()
+
+    def _check_tag(self):
+        """
+        Validates if the instance exists
+        """
+        if not super()._check_tag():
+            return False
+        
+        # Validates if the instance exists
+        if not self._check_instance():
+            return False
+        
+        return True
+
+
+class UserObjectTag(ModelObjectTag):
+    """
+    Object tags used on taxonomies associated with user model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = get_user_model()
+
+
+class LanguageObjectTag(ClosedSystemObjectTag):
+    """
+    Object tag for Languages
+
+    The tags are filtered and validated taking into account the 
+    languages available in Django LANGUAGES settings var
+    """
+
+    system_defined_taxonomy_id = SystemDefinedIds.LanguageTaxonomy.value
+
+    class Meta:
+        proxy = True
+
+    @classmethod
+    def get_tags_query_set(cls, taxonomy: Taxonomy) -> models.QuerySet:
+        """
+        Returns a query set of available languages tags.
+        """
+        available_langs = cls._get_available_languages()
+        return taxonomy.tag_set.filter(external_id__in=available_langs)
```

Comment:
If we make this a classmethod on ObjectTag too, then we can call it in `Taxonomy.get_tags` without have to check whether or not the method exists on the subclass.

So in ObjectTag, add:

```python
    @classmethod
    def get_tags_query_set(cls, taxonomy: Taxonomy) -> models.QuerySet:
        """
        Returns a base query set of the tags available on the given taxonomy.
        
        Subclasses should override this method if they need to provide their own custom tag sets.
        """
        return taxonomy.tag_set
```

And so Taxonomy's `get_tags` method can be updated as you recommended:

```python
    def get_tags(self) -> List[Tag]:
        """
        Returns a list of all Tags in the current taxonomy, from the root(s) down to TAXONOMY_MAX_DEPTH tags, in tree order.

        Annotates each returned Tag with its ``depth`` in the tree (starting at 0).

        Performance note: may perform as many as TAXONOMY_MAX_DEPTH select queries.
        """
        tags = []
        if self.allow_free_text:
            return tags

        ObjectTagClass = self.object_tag_class or ObjectTag
        tag_set = ObjectTagClass.get_tags_query_set(self).prefetch_related("parent")
        parents = None
        for depth in range(TAXONOMY_MAX_DEPTH):
            filtered_tags = tag_set
```

Reply:
Edit: Since we can now subclass Taxonomies, you can just make your custom Taxonomy override get_tags, to set the base queryset.

Reply:
Updated here https://github.com/open-craft/openedx-learning/pull/3/commits/62501cddbab4ed5b2360be061acbeb9b56436fff